### ISSUE TITLE
Lexer, parser, etc, should be disposable and dispose of its memory

### ISSUE THREAD

**Issue by mwpowellhtx** [antlr/antlr4#2482]
I am writing some unit tests using *xunit*, *Visual Studio Runner* (*ReSharper*), and combinatorials. I am finding that the memory in the actual parse verification is never released back to the memory pool. Consequently, I find that the test piles up GB upon GB of memory during the unit test.

My test fixture has these for support:

```C#
private static IEnumerable<Tuple<object, Type>> GetArgument<T>(T obj)
{
    yield return Tuple.Create((object) obj, typeof(T));
}

private static TResult Construct<TResult>(params Tuple<object, Type>[] args)
{
    var ctor = typeof(TResult).GetConstructor(args.Select(x => x.Item2).ToArray());
    Assert.NotNull(ctor);

    var result = ctor.Invoke(args.Select(x => x.Item1).ToArray());
    Assert.NotNull(result);
    Assert.IsType<TResult>(result);

    return (TResult) result;
}
```

And the corner stone:

```C#
protected delegate IParseTree GetWalkableContextCallback(TParser parser);

protected void VerifyParse(string given, GetWalkableContextCallback callback)
{
    VerifyGiven(given);

    var source = Construct<TSource>(GetArgument<ICharStream>(
        Construct<AntlrInputStream>(GetArgument(given).ToArray())
    ).ToArray());

    var parser = Construct<TParser>(GetArgument<ITokenStream>(
        Construct<CommonTokenStream>(GetArgument<ITokenSource>(source).ToArray())
    ).ToArray());

    var walker = Construct<ParseTreeWalker>();
    var listener = Construct<TListener>();

    var context = callback(parser);
    Assert.NotNull(context);

    walker.Walk(listener, context);

    VerifyListener(listener);
}

protected abstract void VerifyListener(TListener listener);
```

But for the ctor and arg checks, allowing for xunit verification, nothing fancy here.

One thing I did find was the *ANTLR4* elements are not disposable. Which if memory is somehow piling up, not getting released for whatever reason, that's a problem I think.

And in usage:

```C#
[Theory, CombinatorialData]
public void OptionStatementShouldParseCorrectly(
    [FullIdentValues] int prefixLength, [FullIdentValues] int prefixParts
    , [FullIdentValues] int suffixLength, [FullIdentValues(0)] int suffixParts
    , [HeterogeneousConstantValues] IConstant expectedConstant)
{
    void TestContext()
    {
        Assert.NotNull(expectedConstant);

        ExpectedProto = new ProtoDescriptor
        {
            Syntax = new SyntaxStatement { },
            Items =
            {
                new OptionStatement
                {
                    Name =
                    {
                        NamePrefix = GetFullIdent(prefixLength, prefixParts),
                        NameSuffix = GetFullIdent(suffixLength, suffixParts)
                    },
                    Value = expectedConstant
                }
            }
        };

        var given = ExpectedProto.ToDescriptorString();

        VerifyParse(given, ctx => ctx.proto());
    }

    using (var context = new DisposableContext(TestContext))
    {
    }
}
```

``FullIdentValuesAttribute`` seeds the set *[1, 3]*, with option to furnish additional values. ``HeterogeneousConstantValuesAttribute`` furnishes my *Protocol Buffer* ``IConstant`` instances for the range of possible constants, from ``IConstant<bool>``, ``IConstant<string>``, ``IConstant<long>``, ``IConstant<double>``.

In all I counted something like 500+ test cases. Nothing that tremendous. However, invoking the *ANTLR4* is extremely inefficient memory management wise.

``DisposableContext`` is my attempt to initiate a ``GC.Collect(GC.MaxGeneration)`` followed by ``GC.WaitForPendingFinalizers()``, to no avail.

And, when my physical memory is consumed, I run into page locks, etc, pushing through virtual memory, etc. Obviously, if the memory issue can be arrested and addressed, that's my preference.

I will add that I am synthesizing a *C#* AST given my *Protov2Listener*. At strategic steps entering (synthesizing) and exiting (resolving, sometimes parsing literals, etc) rules during the listener. In and of itself, that works pretty well as long as the enter/exit chain is properly connected.

Internally, I keep track of the AST state using a stack of sorts.

```C#
public ProtoDescriptor ActualProto { get; protected set; }

protected internal IList<object> Descriptors { get; } = new List<object>();

protected delegate TCurrent SynthesizeCallback<in TContext, out TCurrent>(TContext context)
    where TContext : RuleContext;

protected void OnEnterSynthesizeAttribute<TContext, TCurrent>(TContext context, SynthesizeCallback<TContext, TCurrent> synthesize)
    where TContext : RuleContext
{
    Descriptors.Add(synthesize(context));
}

protected delegate void CollapseToPreviousCallback<TPrevious, in TCurrent>(ref TPrevious previous, TCurrent current);

protected void OnExitResolveSynthesizedAttribute<TPrevious, TCurrent>(
    CollapseToPreviousCallback<TPrevious, TCurrent> callback)
{
    // Short circuit when we fail to match the types.
    var d = Descriptors;

    // ReSharper disable once InvertIf
    if (d.NextToLast() is TPrevious previous
        && d.Last() is TCurrent current)
    {
        callback(ref previous, current);

        d.TryRemoveAt(d.Count - 1);

        /* Whatever work was done to Previous, needs to be replace.
         * This is especially true when working with integral types. */
        d[d.Count - 1] = previous;
    }
}
```

**Comment by mwpowellhtx** [comment]
I suppose it is also possible that the runner and/or *xunit* itself are maintaining some degree of the test fixture instance, *combinatorial* state, etc, however, I should think that when ``VerifyParse`` has completed, that any memory associated with it also get released.

**Comment by sharwell** [comment]
I'll reply more tomorrow, but as a quick note make sure to clear `lexer.Atn` when you clear the one for the parser.

**Comment by mwpowellhtx** [comment]
@sharwell I am satisfied that the issue was "mine", so to speak. I had a lexer rule that was semi- if not fully-recursive in terms of lexer rules. So when I requested the top level ``.proto()`` context, *ANTLR4* faithfully tried to do what I asked it to do, which resulted in churning and chewing up GB upon GB of memory. I resolved this by simplifying the lexer rules to bite size atomic nibblets, refactoring a couple of them to proper parser rules. Literally, thinking instead of lexer rules as being vocabulary and grammar rules being, well, grammatical, in nature. Sorry for the performance confusion. That said... I still think it would be handy for lexer, parser, etc, to be disposable. That would make some sense in terms of their usage when services call them up to process. The performance issue here, however, I think is properly addressed as an error between the keyboard and the chair, if you will. Cheers! :+1:
